thrimshim/thrimshim/main.py

```diff
@@ -36,6 +40,53 @@ def _start_response(status, headers, exc_info=None):
 	return handle
 
 
+
+def authenticate(f):
+	"""Authenticate a token against the database.
+
+	Reference: https://developers.google.com/identity/sign-in/web/backend-auth"""
+	@wraps(f)
+	def decorated_function(*args, **kwargs):
+		if flask.request.method == 'POST':
+			if app.no_authentication:
+				return f(*args, editor='NOT_AUTH', **kwargs)
+
+			userToken = flask.request.json['token']
+			# check whether token is valid
+			try:
+				idinfo = id_token.verify_oauth2_token(userToken, requests.Request(), None)
+				if idinfo['iss'] not in ['accounts.google.com', 'https://accounts.google.com']:
+					raise ValueError('Wrong issuer.')
+			except ValueError:
+				return 'Invalid token. Access denied.', 403 
+
+			# check whether user is in the database
+			email = idinfo['email']
```

Comment:
i'm a little worried about issues around multiple versions of the same email, eg. uppercase, or with extra dots in it. But I guess the version we get from google will always be the same, and we just need to ensure that the same version is the one in the database. Though a case-sensitive match would probably catch most of it.

Reply:
Do you mean case insensitive match?

Reply:
yes, whoops